packages/babel-parser/src/plugins/typescript/index.js

```diff
@@ -1617,18 +1617,24 @@ export default (superClass: Class<Parser>): Class<Parser> =>
           break;
 
         case "module":
-          if (next) this.next();
-          if (this.match(tt.string)) {
-            return this.tsParseAmbientExternalModuleDeclaration(node);
-          } else if (this.tsCheckLineTerminatorAndMatch(tt.name, next)) {
-            return this.tsParseModuleOrNamespaceDeclaration(node);
+          if (!next || !this.hasFollowingLineBreak()) {
+            if (next) this.next();
+            if (!this.isLineTerminator()) {
+              if (this.match(tt.string)) {
+                return this.tsParseAmbientExternalModuleDeclaration(node);
+              } else if (this.match(tt.name)) {
+                return this.tsParseModuleOrNamespaceDeclaration(node);
+              }
+            }
           }
           break;
 
         case "namespace":
-          if (this.tsCheckLineTerminatorAndMatch(tt.name, next)) {
+          if (!next || !this.hasFollowingLineBreak()) {
             if (next) this.next();
-            return this.tsParseModuleOrNamespaceDeclaration(node);
+            if (!this.isLineTerminator() && this.match(tt.name)) {
+              return this.tsParseModuleOrNamespaceDeclaration(node);
+            }
```

Comment:
Similarly, we can avoid checking the line break twice when `next` is `true` (even if the code _seems_ longer because it's less compact:

```suggestion
          // "namespace" cannot be followed by a line break.
          if (next) {
            if (this.hasFollowingLineBreak()) break;
            this.next();
          } else if (this.isLineTerminator()) {
            break;
          }

          if (this.match(tt.name)) {
            return this.tsParseModuleOrNamespaceDeclaration(node);
          }
```
```